server/worker/worker.go

```diff
@@ -0,0 +1,110 @@
+package worker
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+const maxRetries = 5
+
+// Job defines an interface for jobs that can be run by the Worker
+type Job interface {
+	// Name is the unique name of the job.
+	Name() string
+
+	// Run performs the actual work.
+	Run(ctx context.Context, args map[string]interface{}) error
+}
+
+// Worker runs jobs. NOT SAFE FOR CONCURRENT USE.
+type Worker struct {
+	ds  fleet.Datastore
+	log kitlog.Logger
+
+	registry map[string]Job
+}
+
+func NewWorker(ds fleet.Datastore, log kitlog.Logger) *Worker {
+	return &Worker{
+		ds:       ds,
+		log:      log,
+		registry: make(map[string]Job),
+	}
+}
+
+func (w *Worker) Register(jobs ...Job) {
+	for _, j := range jobs {
+		w.registry[j.Name()] = j
```

Comment:
It should probably panic if a job is already registered, as that would be a programming error and registration should happen at app startup anyway, so better to fail fast.

Reply:
I agree